Issue #123 - Deprecate Pre2P

andrewnguyen22 opened the pull request:
## Description

Deprecates the Pre2P module and move it to P2P by closing #123.

## Type of change
Please mark the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [X] Code cleanup

## How Has This Been Tested?
- `make test_all`
- `LocalNet`

## Checklist:
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] No new warnings are generated by my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

---

Co-authored-by: Andrew Nguyen <andrewnguyen@Andrews-MacBook-Pro-2.local>
Co-authored-by: Daniel Olshansky <olshansky.daniel@gmail.com>

Commit:
issue-#123

Commit:
cleanup layer

Olshansk commented on shared/node.go:
```diff
@@ -39,8 +39,8 @@ func Create(cfg *config.Config) (n *Node, err error) {
 
-	// TODO(derrandz): Deprecate `p2p` and replace `pre2p` into its place
-	pre2pMod, err := pre2p.Create(cfg)
+	// TODO(derrandz): Deprecate `p2p` and replace `p2p` into its place
```
Remove comment?

Olshansk commented on p2p/types/network.go:
```diff
@@ -0,0 +1,79 @@
+	"sort"
+)
+
+// CLEANUP(olshansky): See if we can deprecate one of these structures.
```
Could you take a look at this and see if we can simply the code?

Olshansk commented on p2p/types/network.go:
```diff
@@ -0,0 +1,79 @@
+
+// TODO (Team) break off these types into diff files
+
+func (ab *AddrList) Find(address string) (index int, found bool) {
```
s/ab/al

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -4,42 +4,44 @@ import (
 	"encoding/binary"
 	"encoding/hex"
 	"fmt"
+	types2 "github.com/pokt-network/pocket/p2p/types"
```
Can you name this `typesP2P`?

Olshansk replied:
Done.

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -4,42 +4,44 @@ import (
-	addrList     []string
-	maxNumLevels uint32
+	addrBookMap            types2.AddrBookMap
+	addrList               []string
```
We defined a type for `AddrList` in the other file. Please use it

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -50,39 +52,64 @@ func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook)
+		if n.redundancyLayerEnabled {
+			// redundancy layer is simply one final send to the original +1/3 && -1/3
+			level = n.maxNumLevels
+			// ensure not an echo-chamber
```
How does level=-1 ensure this?

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -92,8 +119,8 @@ func (n *rainTreeNetwork) networkBroadcastAtLevel(data []byte, level uint32, non
-func (n *rainTreeNetwork) demote(rainTreeMsg *typesPre2P.RainTreeMessage) error {
-	if rainTreeMsg.Level > 0 {
+func (n *rainTreeNetwork) demote(rainTreeMsg *types2.RainTreeMessage) error {
+	if rainTreeMsg.Level >= 0 {
```
Why do we demote at level 0?

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -50,39 +52,64 @@ func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook)
 		Nonce: nonce,
 	}
+	// This is handled either by the redundancy layer
+	if level == 0 {
```
For clarity of code, can you move `redundancyLayer`  and `cleanupLayer` into separate functions. Similar to how `demote` is a separate function right now.

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -50,39 +52,64 @@ func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook)
+			// ensure not an echo-chamber
+			msg.Level = -1
+		} else {
+			if err := n.demote(msg); err != nil {
```
I don't think we need to call `demote`  if level == 0?

andrewnguyen22 replied:
Cleanup layer is -1

Olshansk commented on p2p/raintree/network.go:
```diff
@@ -50,39 +52,64 @@ func NewRainTreeNetwork(addr cryptoPocket.Address, addrBook typesPre2P.AddrBook)
-		if err = n.networkSendInternal(bz, addr1); err != nil {
-			log.Println("Error sending to peer during broadcast: ", err)
-		}
+	if addr1 == nil {
```
I think we'll need to revert it to the previous if structure if we move cleanup into a separate function.

I understand that both approaches work the exact same, and have the same performance, but I tink that doing it the other way will provide a clearer path of how the code operates, even if it may be more redundant.

E.g.

// Cleanup
if level == -1:
 callCLeanupLayer()
 return

// Normal Left 
 if addr1 = getFirstTargets() && addr1 != nil:
  SendToAddr1
  
// Normal right
if addr1 = getFirstTargets() && addr1 != nil:
  SendToAddr1

Olshansk commented on p2p/raintree/addrbook_utils_test.go:
```diff
@@ -172,18 +179,33 @@ func testRainTreeMessageTargets(t *testing.T, expectedMsgProp *ExpectedRainTreeM
+			level = getMaxAddrBookLevels(addrBook)
+		}
+		if level == -1 {
+			addr1, addr2, _ = getLeftAndRight(addrList, addrBookMap)
```
Use `ok` and assert on `true`

Olshansk commented on p2p/raintree/addrbook_utils_test.go:
```diff
@@ -161,6 +166,8 @@ func TestRainTreeAddrBookTargetsTwentySevenNodes(t *testing.T) {
 		{3, "X", "F"},
 		{2, "U", "["},
 		{1, "S", "W"},
+		{0, "X", "F"},  // redundancy
```
Requests (for all the tree diagrams):
1. Could you update the spacing. I messed up in a few places
2. I think adding ** around where the expected raintree target is in the tree will help (e.g. around *U* and around *[*)
3. Could you add redundancy and cleanup via some visual format as well. I know we discussed this before, but even I forgot and am having a tough time trying to understand how it works.

andrewnguyen22 commented on p2p/raintree/addrbook_utils.go:
```diff
@@ -70,30 +57,48 @@ func (n *rainTreeNetwork) getTarget(level uint32, targetPercentage float64) (cry
 // the addr book was updated), but we're using `maxNumLevels` associated with the number of
 // validators at the current height.
-func (n *rainTreeNetwork) getAddrBookLengthAtHeight(level uint32) int {
+func (n *rainTreeNetwork) getAddrBookLengthAtHeight(level int32) int {
```
Cleanup layer uses -1

andrewnguyen22 commented on p2p/CHANGELOG.md:
```diff
@@ -0,0 +1,13 @@
+
+## [0.0.0.0] - 2022-07-26
+
+- Raintree first iteration (no redundancy or cleanup layer)
```
Yes, will fix

Commit:
fixed pr changes

andrewnguyen22 commented:
@Olshansk I fixed all of your asks.

I'm throwing this to you because I have no idea how to work with / understand the rain tree module level tests (which are now failling).

My asks are:
- Fix the tests with the redundancy layer (use my visual diagram and cleaned up code to understand)
- Create an issue to make these tests more straightforward
- Take some time offline to explain to me how these work

Olshansk commented:
@andrewnguyen22 If we split out redundancy + cleanup into a separate PR from Deprecate Pre2P, then we can deprecate Pre2P without needing to rush the improvement of the tests. I realize this requires extra work, but let me know if you'd rather us have:

1. Focus on fixing the tests and improving their understanding
2. Splitting it out in two independent PRs (1 deprecate Pre2P, and then add clean/redundancy)

Alternatively (hopefully a group fix), we can take away the lesson to not group cleanup & logic layer changes and go over this in protocol hour tomorrow?

Olshansk committed:
Minor edits to Makefile

Olshansk committed:
Remove redundancy & cleanup layer additions

Olshansk committed:
Update changelog
